### DISCUSSION TITLE
Connect to two bluetooth devices

### DISCUSSION THREAD

**Discussion by CB925** [pybricks/support#1816] [Q&A]
Hi all. I found instructions about how you can use Pybricks to connect to the duplo train. It works, however when I connect to the train it's not possible to connect to the lego remote control. And if I connect to the remote first, it's not possible to connect to the train later. I get an error message about the system being busy on the second connection attempt. Can the hub only connect to one bluetooth device at the same time, or something else wrong? I'm using the technic hub.

**Comment by laurensvalk** [comment]
> Can the hub only connect to one bluetooth device at the same time

Yes, this is correct. From the Technic Hub's point of view, connecting to the Duplo Train is very similar to connecting to the Remote, and it can do only one of those. (After all, it also still connects to your computer, so the connectivity all adds up and there is a limit.)

Just as an alternate idea, could you creatively make the Technic Hub itself "be" the remote, by attaching handles or sensors to it?

**Reply by yoredale** [reply]
This is a shame, I'd just been trying to implement this too, and had run into the same issue - May have over-promised my 4 year old!

Does this limitation also apply to controlling multiple trains with one hub? My next step was to implement 2 levers, one for each train.

Is this a limitation of the Technic Hub in particular (ie because of limited resources) or of the hubs in general?

Thanks

**Reply by RaceBo** [reply]
> Yes, this is correct. From the Technic Hub's point of view, connecting to the Duplo Train is very similar to connecting to the Remote, and it can do only one of those. (After all, it also still connects to your computer, so the connectivity all adds up and there is a limit.)

Add me to the list. I actually decided between BrikNil with Raspi and Pybricks on technic hub and chose pybricks mainly because of higher WAF and Battery operation. I was not aware of this - very disappointing - limitation. Is there any chance to get this working manually? As far as I understand, original lego equipment can do this (albeit not with duplo trains). 
Also, your reply did actively mention the technic hub. Does Pybricks support multiple bluetooth devices with any of the other hubs?

Thanks in advance for your answer.

**Reply by laurensvalk** [reply]
>  As far as I understand, original lego equipment can do this (albeit not with duplo trains).

With the LEGO apps, the Technic hub can only do a single connection -- everything happens through your phone. Even the remote goes through your phone, so your phone is always in the loop.

With Pybricks, you can simultaneously be connected to your PC (or phone), a remote (or train), and even communicate with other hubs as well through broadcast messaging.

So Pybricks lets you have 3 simultaneous wireless activities instead of just one, which is about on the limit of what the Technic hub can physically do. (With the wireless communication between hubs, the possibilities should be endless since this can go far beyond 3 or more hubs.)

How about using the Technic Hub as a remote for the train like in [this example](https://pybricks.com/project/control-the-duplo-train/)? You can add levers, use the hub button, etc.

**Reply by yoredale** [reply]
I have a third Technic Hub arriving today to implement 1 remote to 2 duplo trains with 1 hub for the remote, and 1 hub for each train, using inter-hub communication to send the commands from the remote hub to the controller hubs.

I added a function to duplo.py that will rename a train, and altered the init function to accept a name argument, so I could rename the trains and subsequently connect a hub to a specific train. I did experiment with connecting both trains to 1 hub, but it fails when you try to connect the second train.

So what I'm going to end up with is 2 lever controllers like Laurens example that connect to specific trains, with a train remote that can also control both trains. To add to the fun, the control levers will also move to the position set by the remote.

**Comment by CB925** [comment]
I haven't tried this yet, but I guess I can get this to work with two hubs? One hub connected to the train, and the other one to the remote. So can the remote-hub send instructions to the train-hub when a button is pressed.

**Reply by laurensvalk** [reply]
Yes, you can do that. Here's an example that gets quite close to that: https://pybricks.com/project/technic-42146-powered-up-remote/ You'd just need to combine the receiving program for this with the Duplo control example.

It is admittedly quite an elaborate setup to control a train. Maybe you can also use the hubs for some automated switches or automated barriers? One of them might as well be on board too :smile: .

**Reply by laurensvalk** [reply]
This should be quite quick and smooth once up and running (even if it may appear a bit slow while the hubs are also still connected to the PC).

**Reply by RaceBo** [reply]
Well, yes, you can do that, of cause. But for my setup it means I need 4 hubs (1 for each train and one for each remote), plus the remotes itself. That's an invest of around 200 Euros plus time plus storage - just to remote control the trains without adding up screen time for my son. 
Your answer creates a glaze of hope in me: I've already tried to cut PC connection, hoping that 2 Bluetooth connections work at least. But they don't. Is there a chance, if it is only a problem of resources on the hub? How does BrickNil do it with a single Bluetooth Adapter?

**Reply by laurensvalk** [reply]
For what it's worth, if the goal is to reduce screen time (which was also my goal), my kids really enjoyed a [setup like this].(https://pybricks.com/project/control-the-duplo-train/), which is a lot simpler than a remote with 7 buttons. You can even get a bit more creative and make it a "tilt to drive" setup so you don't even have to build anything. 

> How does BrickNil do it with a single Bluetooth Adapter?

In that scenario each hub or remote has only one connection (each to the computer).
